python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,470 @@
+
+        >>> block_1 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [1.0, 2.0, 4.0],
+        ...             [3.0, 5.0, 6.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[0], [1], [2]])),
+        ... )
+        >>> block_2 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [5.0, 8.0, 2.0],
+        ...             [1.0, 2.0, 8.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[3], [4], [5]])),
+        ... )
+        >>> keys = Labels(names=["key"], values=np.array([[0], [1]]))
+        >>> tensor = TensorMap(keys, [block_1, block_2])
+
+        Create modules
+
+        >>> linear = torch.nn.Linear(3, 1, bias=False)
+        >>> with torch.no_grad():
+        ...     _ = linear.weight.copy_(torch.tensor([1.0, 1.0, 1.0]))
+        ...
+        >>> modules = [linear, deepcopy(linear)]
+        >>> # you could also extend the module by some nonlinear activation function
+
+        Create ModuleMap from this ModucDict and apply it
+
+        >>> module_map = ModuleMap(tensor.keys, modules)
+        >>> out = module_map(tensor)
+        >>> out
+        TensorMap with 2 blocks
+        keys: key
+               0
+               1
+        >>> out[0].values
+        tensor([[ 7.],
+                [14.]], grad_fn=<MmBackward0>)
+        >>> out[1].values
+        tensor([[15.],
+                [11.]], grad_fn=<MmBackward0>)
+
+        Lets look at the metadata
+
+        >>> tensor[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (3): ['properties']
+            gradients: None
+        >>> out[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (1): ['_']
+            gradients: None
+
+        It got completely lost because we cannot know in general what the output is.
+        You can add in the initialization of the ModuleMap a TensorMap that contains
+        the intended output Labels.
+    """
+
+    def __init__(
+        self,
+        in_keys: Labels,
+        modules: Sequence[Module],
+        out_tensor: Optional[TensorMap] = None,
```

Comment:
It seems that the only reason to have the `out_tensor` in the ModuleMap is to provide a name for properties dimension of the output tensor. In my experience, this feature is really redundant, because a simple `feature` or `_` or any other placeholder works perfectly, since you never explicitly address this name anyway. 

Moreover, usually the presence of the output tensor or array means that the values will be stored in this specific tensor, and this is not the case, since we actually create the new `TensorMap`.

Reply:
How about `out_features=None`, and have `None` map to `Labels.range("_", len(output))`?

Reply:
I think giving the option for the end user to specify properties labels if they want to is good (defaulting to `Labels.range("_", len(features))`).

But agreed on the implication that data would be stored there. How about `out_metadata`? or `out_properties: Tuple[Labels, List[Labels]]` (as a poor man's `Dict[LabelEntry, Labels]`)?

Reply:
> I think giving the option for the end user to specify properties labels if they want to is good (defaulting to `Labels.range("_", len(features))`).
> 
> But agreed on the implication that data would be stored there. How about `out_metadata`? or `out_properties: Tuple[Labels, List[Labels]]` (as a poor man's `Dict[LabelEntry, Labels]`)?

I think that the general issue here is the following
 - If we want to set the properties dimension name for hidden layers, we can't really say much about the nature of this data, except that this is a hidden representation. Thus the hardcoded placeholder name is a good choice IMO
 - If we want to set the name for a last layer only, then it's not really necessary, since we know what type of property we expect to get
 
So the only really useful case in this situation will be a multi-modality model (for instance which predicts both absolute energy and the energy of formation), and we store the output in format of a single tensormap with two different output blocks corresponding to each output type each.

Reply:
> If we want to set the name for a last layer only, then it's not really necessary, since we know what type of property we expect to get

Since the ModuleMap can allow arbitrary modules, there will be modules that perform operations with tractable property names. You can say the module developer needs to do this in their module implementation, but for that you need to override the forward function around the ModuleMap. Even you create a new Tensor.Map,  since you cannot change the Labels after creation, or you copy the whole forward function from ModulMap  to add the change of the label. I feel like this is more convenient to do. And for the some output module you even said that it is useful to have. I am not sure if in practice people use it, but in principle I can see some useful applications for it.